drivers/axi_core/spi_engine/spi_engine.c

```diff
@@ -0,0 +1,958 @@
+	spi_engine_set_transfer_wdith(local_desc, spi_engine_init->data_width);
+
+	/* Get current data width */
+	spi_engine_read(eng_desc, SPI_ENGINE_REG_VERSION, &spi_engine_version);
+
+	*desc = local_desc;
+
+	dev_info(desc, "Spi engine v%d.%d.%d succesfully initialized at 0x%X.",
+		 (spi_engine_version >> 16),
+		 ((spi_engine_version >> 8) & 0xFF),
+		 (spi_engine_version & 0xFF),
+		 desc);
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Write/read on the spi interface
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param data Pointer to data buffer
+ * @param bytes_number Number of bytes to transfer
+ * @return int32_t - SUCCESS if the transfer finished
+ *		   - FAILURE if the memory allocation or transfer failed
+ */
+int32_t spi_engine_write_and_read(struct spi_desc *desc,
+				  uint8_t *data,
+				  uint8_t bytes_number)
+{
+	uint8_t 		i;
+	uint8_t 		word_len;
+	uint8_t 		words_number;
+	int32_t 		ret;
+	struct spi_engine_msg	*msg;
+	struct spi_engine_desc	*desc_extra;
+
+	desc_extra = desc->extra;
+
+	words_number = spi_get_words_number(desc_extra, bytes_number);
+
+	msg = (spi_engine_msg *)malloc(sizeof(*msg));
+	msg->cmds = (spi_engine_cmd_queue*)malloc(sizeof(*msg->cmds));
+	if (!msg->cmds || !msg)
+		return FAILURE;
+
+	msg->tx_buf =(uint32_t*)malloc(words_number * sizeof(msg->tx_buf[0]));
+	msg->rx_buf =(uint32_t*)malloc(words_number * sizeof(msg->rx_buf[0]));
+	msg->length = words_number;
+
+	/* Init the rx and tx buffers with 0s */
+	for (i = 0; i < words_number; i++) {
+		msg->tx_buf[i] = 0;
+		msg->rx_buf[i] = 0;
+	}
+
+	/* Get the length of transfered word */
+	word_len = spi_get_word_lenght(desc_extra);
+
+	/* Make sure the CS is HIGH before starting a transaction */
+	msg->cmds->next = NULL;
+	msg->cmds->cmd = CS_HIGH;
+	spi_engine_queue_add_cmd(&msg->cmds, CS_LOW);
+	spi_engine_queue_add_cmd(&msg->cmds, WRITE_READ(bytes_number));
+	spi_engine_queue_add_cmd(&msg->cmds, CS_HIGH);
+
+	/* Pack the bytes into engine WORDS */
+	for (i = 0; i < bytes_number; i++)
+		msg->tx_buf[i / word_len] |= data[i] << (desc_extra->data_width-
+					     (i % word_len + 1) * 8);
+
+	ret = spi_engine_transfer_message(desc, msg);
+
+	/* Skip the first byte ( dummy read byte ) */
+	for (i = 1; i < bytes_number; i++)
+		data[i - 1] = msg->rx_buf[(i) / word_len] >>
+			      (desc_extra->data_width -
+			      ((i) % word_len + 1) * 8);
+
+	spi_engine_queue_free(&msg->cmds);
+	free(msg->tx_buf);
+	free(msg->rx_buf);
+	free(msg);
+	return ret;
+}
+
+/**
+ * @brief Initialize the SPI engine's offload module
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param param Structure containing the offload init parameters
+ * @return int32_t This function allways returns SUCCESS
+ */
+int32_t spi_engine_offload_init(struct spi_desc *desc,
+				const struct spi_engine_offload_init_param *param)
+{
+	struct spi_engine_desc	*eng_desc;
+	struct axi_dmac_init	*dmac_init;
```

Comment:
I would rather pass them as pointers and allocate the memory inside the spi engine driver instead of constraining the user to allocate memory in his application.

Reply:
Sorry, I've missed that init